File: 07_linked_list_and_arrays/4_trapping_rainwater.py
class Solution:
    def trap(self, heights: List[int]) -> int:
        n = len(heights)

        pre_max = []
        suff_max = []

        curr_pre_max = curr_suff_max = 0

        for i in range(n):
            curr_pre_max = max(curr_pre_max, heights[i])
            pre_max.append(curr_pre_max)

            curr_suff_max = max(curr_suff_max, heights[n - i - 1])
            suff_max.insert(0, curr_suff_max)

        res = 0
        for a, b, c in zip(heights, pre_max, suff_max):
            res += min(b, c) - a

        return res

File: 07_linked_list_and_arrays/4_trapping_rainwater.rs
use std::cmp::{max, min};

impl Solution {
    pub fn trap(heights: Vec<i32>) -> i32 {
        let (mut pmax, mut smax) = (0, 0);
        let mut pmax_vec: Vec<i32> = vec![];
        let mut smax_vec: Vec<i32> = vec![];
        let n = heights.len();

        for (i, &height) in heights.iter().enumerate() {
            pmax = max(pmax, height);
            pmax_vec.push(pmax);

            smax = max(smax, heights[n - i - 1]);
            smax_vec.push(smax);
        }

        let mut res = 0;
        smax_vec.reverse();
        for i in 0..n {
            res += min(pmax_vec[i], smax_vec[i]) - heights[i];
        }

        res
    }
}
